pkg/v3/coordinator/conditional_report_coordinator.go

```diff
@@ -0,0 +1,342 @@
+	idBlocks       *util.Cache[idBlocker] // should clear out when the next perform with this id occurs
+	activeKeys     *util.Cache[bool]
+	cacheCleaner   *util.IntervalCacheCleaner[bool]
+	idCacheCleaner *util.IntervalCacheCleaner[idBlocker]
+
+	// configurations
+	minConfs int
+
+	// run state data
+	running atomic.Bool
+	chStop  chan struct{}
+}
+
+// NewConditionalReportCoordinator provides a new conditional report coordinator. The coordinator
+// should be started before use.
+func NewConditionalReportCoordinator(
+	events EventProvider,
+	minConfs int,
+	logger *log.Logger,
+	encoder Encoder,
+) *conditionalReportCoordinator {
+	c := &conditionalReportCoordinator{
+		logger:         logger,
+		events:         events,
+		minConfs:       minConfs,
+		idBlocks:       util.NewCache[idBlocker](DefaultLockoutWindow),
+		activeKeys:     util.NewCache[bool](time.Hour), // 1 hour allows the cleanup routine to clear stale data
+		idCacheCleaner: util.NewIntervalCacheCleaner[idBlocker](DefaultCacheClean),
+		cacheCleaner:   util.NewIntervalCacheCleaner[bool](DefaultCacheClean),
+		chStop:         make(chan struct{}, 1),
+		encoder:        encoder,
+	}
+
+	return c
+}
+
+// isPending returns true if a key should be filtered out.
+func (rc *conditionalReportCoordinator) isPending(key ocr2keepers.UpkeepPayload) bool {
+	blockKey := ocr2keepers.BlockKey(strconv.FormatInt(key.Trigger.BlockNumber, 10))
+
+	// only apply filter if key id is registered in the cache
+	if bl, ok := rc.idBlocks.Get(string(key.Upkeep.ID)); ok {
+		isAfter, err := rc.encoder.After(blockKey, bl.TransmitBlockNumber)
+		if err != nil {
+			return true
+		}
+
+		// do not filter the key out if key block is after block in cache
+		return !isAfter
+	}
+
+	return false
+}
+
+// Accept sets the pending status for a key
+func (rc *conditionalReportCoordinator) Accept(key ocr2keepers.ReportedUpkeep) error {
+	blockKey := ocr2keepers.BlockKey(strconv.FormatInt(key.Trigger.BlockNumber, 10))
+
+	// If a key is already active then don't update filters, but also not throw errors as
+	// there might be other keys in the same report which can get accepted
+	if _, ok := rc.activeKeys.Get(string(key.UpkeepID)); !ok {
+		// Set the key as accepted within activeKeys
+		rc.activeKeys.Set(string(key.UpkeepID), false, util.DefaultCacheExpiration)
+
+		// Set idBlocks with the key as checkBlockNumber and IndefiniteBlockingKey as TransmitBlockNumber
+		rc.updateIdBlock(string(key.UpkeepID), idBlocker{
+			CheckBlockNumber:    blockKey,
+			TransmitBlockNumber: IndefiniteBlockingKey,
+		})
+	}
+
+	return nil
+}
+
+// IsTransmissionConfirmed returns whether the upkeep was successfully
+// completed or not
+func (rc *conditionalReportCoordinator) IsTransmissionConfirmed(key ocr2keepers.UpkeepPayload) bool {
+	// key is confirmed if it both exists and has been confirmed by the log
+	// poller
+	confirmed, ok := rc.activeKeys.Get(string(key.Upkeep.ID))
+	return !ok || (ok && confirmed)
```

Comment:
Is this existing functionality correct - if the upkeep ID doesn't exist in `activeKeys` we consider it completed?

Reply:
Yes. The active keys has a timeout, but once the key is gone we assume that it is completed. This could happen if a transaction is stuck in the mempool for a long period of time. We then assume completion and allow further checks and performs of the upkeep. If the stuck transaction does get transmitted after a more up-to-date transaction does, the old one will result in a stale event. This results in a bit of gas waste in the scenario, but is functional.